compiler/rustc_mir/src/transform/validate.rs

```diff
@@ -68,79 +69,24 @@ pub fn equal_up_to_regions(
         return true;
     }
 
-    struct LifetimeIgnoreRelation<'tcx> {
-        tcx: TyCtxt<'tcx>,
-        param_env: ty::ParamEnv<'tcx>,
-    }
-
-    impl TypeRelation<'tcx> for LifetimeIgnoreRelation<'tcx> {
-        fn tcx(&self) -> TyCtxt<'tcx> {
-            self.tcx
-        }
-
-        fn param_env(&self) -> ty::ParamEnv<'tcx> {
-            self.param_env
-        }
-
-        fn tag(&self) -> &'static str {
-            "librustc_mir::transform::validate"
-        }
-
-        fn a_is_expected(&self) -> bool {
-            true
-        }
-
-        fn relate_with_variance<T: Relate<'tcx>>(
-            &mut self,
-            _: ty::Variance,
-            a: T,
-            b: T,
-        ) -> RelateResult<'tcx, T> {
-            // Ignore variance, require types to be exactly the same.
-            self.relate(a, b)
-        }
-
-        fn tys(&mut self, a: Ty<'tcx>, b: Ty<'tcx>) -> RelateResult<'tcx, Ty<'tcx>> {
-            if a == b {
-                // Short-circuit.
-                return Ok(a);
-            }
-            ty::relate::super_relate_tys(self, a, b)
-        }
-
-        fn regions(
-            &mut self,
-            a: ty::Region<'tcx>,
-            _b: ty::Region<'tcx>,
-        ) -> RelateResult<'tcx, ty::Region<'tcx>> {
-            // Ignore regions.
-            Ok(a)
-        }
-
-        fn consts(
-            &mut self,
-            a: &'tcx ty::Const<'tcx>,
-            b: &'tcx ty::Const<'tcx>,
-        ) -> RelateResult<'tcx, &'tcx ty::Const<'tcx>> {
-            ty::relate::super_relate_consts(self, a, b)
-        }
-
-        fn binders<T>(
-            &mut self,
-            a: ty::Binder<T>,
-            b: ty::Binder<T>,
-        ) -> RelateResult<'tcx, ty::Binder<T>>
-        where
-            T: Relate<'tcx>,
-        {
-            self.relate(a.skip_binder(), b.skip_binder())?;
-            Ok(a)
-        }
-    }
-
-    // Instantiate and run relation.
-    let mut relator: LifetimeIgnoreRelation<'tcx> = LifetimeIgnoreRelation { tcx: tcx, param_env };
-    relator.relate(src, dest).is_ok()
+    // Normalize lifetimes away on both sides, then compare.
+    let param_env = param_env.with_reveal_all_normalized(tcx);
+    let normalize = |ty: Ty<'tcx>| {
+        tcx.normalize_erasing_regions(
+            param_env,
+            ty.fold_with(&mut BottomUpFolder {
+                tcx,
+                // We just erase all late-bound lifetimes, but this is not fully correct (FIXME):
```

Comment:
I don't fully understand this comment. I'm not sure how much it matters, but it'd be nice to have a more detailed explanation of how lifetimes could matter. I guess it depends on a bit of context that I seem to be missing, about the broader role of this function.

Reply:
This is the concern that @eddyb keeps raising; [another instance of that is here](https://github.com/rust-lang/rust/pull/77196#issuecomment-719919173). It's something about using lifetimes in invariant positions to force two different trait implementations to be used which could have different associated types which could make this check unsound. I don't have more than a fuzzy understanding of this, and I forgot if we ever were able to come up with a concrete example.

Reply:
Apologies for not providing an example initially, given how trivial it is to construct "transmute in safe code" gadgets from associated type unsoundness (the hard part is you have to intentionally make `rustc` unsound enough to truly confirm something else doesn't defeat the example).
I've just posted https://github.com/rust-lang/rust/pull/77196#issuecomment-721036102, hope that helps.

Reply:
@eddyb also pointed out that we should probably clarify the comment to explicitly mention projections -- specifically, if I understood correctly, the concern is about erasing a lifetime in `T` when the type is `T::Assoc`.

Reply:
I see, so the concern here is that we are not preserving the binding levels? Let me re-read this code then.

Reply:
OK well I definitely agree this requires a FIXME. I don't know that it has to block this PR from *landing*, but I'm not sure I have a good handle on the full repercussions here. I guess the question is why we are not using an existing function to erase or normalize regions and instead rolling our own here.

Reply:
> I see, so the concern here is that we are not preserving the binding levels? Let me re-read this code then.

Binding levels? I have no idea what you are talking about.^^

>  I guess the question is why we are not using an existing function to erase or normalize regions and instead rolling our own here.

I tried all the existing ones (that I could find) and none of them worked.^^ (As in, none of them accepted all the code out there.) Then I asked @eddyb what to do and we ended up with this.

Reply:
> I don't know that it has to block this PR from landing

No, it certainly is not a blocker. This is an old FIXME that was removed when we switched to `TypeRelation` and now re-appears since we are switching back to `BottomUpFolder`.

Reply:
What I meant by "binding level" is that `for<'a> fn(fn(&'a u32))` and `fn(for<'a> fn(&'a u32))` are considered distinct types (and `fn(&'a u32)`, where `'a` appears free, is also distinct) by various parts of Rust. We've been slowly changing the model here but I think we'll not be able to change this completely. In particular, the coherence code accepts distinct impls (as eddyb demonstrated) for some such cases.

Reply:
see also https://github.com/rust-lang/rust/issues/56105 and https://github.com/rust-lang/rust/pull/72493

Reply:
> What I meant by "binding level" is that for<'a> fn(fn(&'a u32)) and fn(for<'a> fn(&'a u32)) are considered distinct types

Sure, it would be really strange if those would be the same type. The outer function in the latter may pick different lifetimes for different calls of the inner function; that is not the case for the former.